language/compiler/ir-to-bytecode/syntax/src/spec_language_ast.rs

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::ast::{BinOp, CopyableVal, StructName, UnaryOp};
+use libra_types::account_address::AccountAddress;
+
+/// AST for the Move Prover specification language. Just postconditions for now
+
+/// A location that can store a value
+#[derive(PartialEq, Debug, Clone)]
+pub enum StorageLocation {
+    /// A formal of the current procedure
+    Formal(String),
+    /// A resource of type `type_` stored in global storage at `address`
+    GlobalResource {
+        type_: StructName,
+        address: Box<StorageLocation>,
+    },
+    /// An access path rooted at `base` with nonempty offsets in `fields`
+    AccessPath {
+        base: Box<StorageLocation>,
+        fields: Vec<String>,
+    },
+    /// Value of a storage location *after* the current procedure executes. Not applicable to
+    /// TxnSenderAddress. Address, or Return
+    Old(Box<StorageLocation>),
+    /// Sender address for the current transaction
+    TxnSenderAddress,
+    /// Account address constant
+    Address(AccountAddress),
```

Comment:
I am puzzled why this is a storage location and not a Constant in SpecExp.

Reply:
Good question. It is actually both at the moment because I was too lazy to separate Move constants from spec constants (though we will eventually want to do this).

However, I think we do need address constants as part of the storage location grammar to be able to say things about resources published at hardcoded addresses (e.g., `ensures global_exists<T>(0xA550C18)`).